### ISSUE TITLE
Preview feature feedback: Integrated seeding 

### ISSUE THREAD

**Issue by 2color** [prisma/prisma#5161]
Please share your feedback about the integrated seeding functionality that was released in [v2.15.0](https://github.com/prisma/prisma/releases/tag/2.15.0) in this issue.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo. 
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.

If you have any questions, don't hesitate to ask them in the [`#prisma-migrate`](https://prisma.slack.com/archives/CBFFY3066) channel in the [Prisma Slack](https://slack.prisma.io).

**Comment by barakcoh** [comment]
Is there a way to run `prisma db seed` when the Prisma folder isn't at its default location?

**Comment by albertoperdomo** [comment]
@barakcoh This is a known issue, tracked [here](https://github.com/prisma/prisma/issues/5213). Most likely to be solved in the next release.

**Comment by rajuashok** [comment]
Just tried this today. I'm getting the following error:
```
Running `ts-node seed.ts` ...
/Users/ashok/src/flixed/api/prisma/seed.ts:37
import { PrismaClient } from '@prisma/client';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

**Comment by rajuashok** [comment]
Changing my `tsconfig.json` to have this:

`"module": "commonjs",`

instead of this:

`"module": "esnext",`

fixed it.

**Comment by pm-dev** [comment]
This script is saying it can't find my seed.ts file:
<img width="738" alt="Screen Shot 2021-01-21 at 9 21 56 PM" src="https://user-images.githubusercontent.com/3067328/105437275-c4816f00-5c2e-11eb-951f-47d3438b8687.png">

**Comment by 2color** [comment]
@pm-dev It looks like you are running the `npx prisma db seed --preview-feature` from the `prisma` folder.

Running it in the root of your repository (where the `package.json` is) should fix that.

**Comment by pm-dev** [comment]
> @pm-dev It looks like you are running the `npx prisma db seed --preview-feature` from the `prisma` folder.
> 
> Running it in the root of your repository (where the `package.json` is) should fix that.

Confirming that did work. (Thought I had tried that). Thanks @2color

**Comment by aruld** [comment]
I am not seeing the data getting seeded in sqlite, I can confirm the data is getting deleted from sqlite. Here is my seed.js script: https://github.com/aruld/server-components-demo/blob/sqlite/prisma/seed.js

<img width="948" alt="Screen Shot 2021-01-23 at 3 00 40 PM" src="https://user-images.githubusercontent.com/442380/105614006-d7f61c80-5d8b-11eb-827f-2d9d6f38090b.png">

any clue what might be happening here?

**Comment by thetw** [comment]
> Running it in the root of your repository (where the `package.json` is) should fix that.

The `seed` command doesn't seem to correctly escape whitespace, because when i run `npx prisma db seed --preview-feature` from the project root it breaks with an incomplete path. The problem seems to be that i've got whitespace in my hard drives name.
```
Running `ts-node seed.ts` ...
Error: Cannot find module '/Volumes/Storage'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
    at main (/Volumes/Storage SSD/Users/thetw/Coding/prisma-test/node_modules/.pnpm/ts-node@9.1.1_typescript@4.1.3/node_modules/ts-node/src/bin.ts:198:14)
    at Object.<anonymous> (/Volumes/Storage SSD/Users/thetw/Coding/prisma-test/node_modules/.pnpm/ts-node@9.1.1_typescript@4.1.3/node_modules/ts-node/src/bin.ts:288:3)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
Error: Command failed with exit code 1: ts-node /Volumes/Storage SSD/Users/thetw/Coding/prisma-test/prisma/seed.ts
```
When i run `ts-node ./prisma/seed.ts` myself with a relative path it indeed works.

**Comment by StErMi** [comment]
> Changing my `tsconfig.json` to have this:
> 
> `"module": "commonjs",`
> 
> instead of this:
> 
> `"module": "esnext",`
> 
> fixed it.

what are the implications of your change? I also get that `seed.ts` error

**Comment by JakubKoralewski** [comment]
> what are the implications of your change? I also get that `seed.ts` error

I was also hesitant; but my Next.js app builds fine, and IDE seems to not throw any errors with this in my project, so I guess it must be fine(?).

---

Another error regarding the `"baseUrl"` config option in `tsconfig.json`:
```
Cannot find module 'lib/prisma/prisma'
```
<details>
<summary>
I'm using `baseUrl` in my `tsconfig.json`
</summary>

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "noImplicitAny": true,
    "esModuleInterop": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "baseUrl": "./src",
    "sourceMap": true
  },
  "exclude": ["node_modules", ".next", "out"],
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", "**/*.js"]
}

```

</details>

 In my custom seeding script I work around this by running my seeding script as follows: `ts-node -r tsconfig-paths/register -P prisma/seeding/tsconfig.script.json prisma/seeding/initiateDb.ts`. I guess I'll stick to my custom script for now. (Note that the `tsconfig.script.json` is not needed anymore as I have just changed my `tsconfig.json` `"compilerOptions"`.`"module"` from `ESNext` to `commonjs`.

Will `tsconfig-paths` be used to support `baseUrl` and other custom `paths`?

**Comment by 2color** [comment]
> any clue what might be happening here?
@aruld 

Make sure to return the Prisma Client calls from the `map`function inside `Promise.all`:

```js
async function main() {
  await prisma.notes.deleteMany();
  const res = await Promise.all(
    seedData.map((data) => {
      const newNote = prisma.notes.create({data}); // needs to be returned
      return newNote;
    })
  );
}
```

**Comment by albertoperdomo** [comment]
@thetw Thanks for reporting! This is indeed a bug and a fix is underway. You can expect this to go live in 2.16.0.

**Comment by Jolg42** [comment]
@JakubKoralewski 
From https://github.com/TypeStrong/ts-node
>ts-node-script is recommended because it enables --script-mode, discovering tsconfig.json relative to the script's location instead of process.cwd(). This makes scripts more portable.

It looks like using `ts-node --script-mode` would be a good option as you could have this setup, what do you think?
- ./prisma/seed.ts
- ./prisma/tsconfig.json

Note: `process.exit(0)` should be used for exiting with success and all non zero code are for exiting with an error. 
Are you using that [seed template](https://www.prisma.io/docs/guides/prisma-guides/seed-database#integrated-seeding-with-prisma-migrate)?

**Comment by JakubKoralewski** [comment]
> > ts-node-script is recommended because it enables --script-mode, discovering tsconfig.json relative to the script's location instead of process.cwd(). This makes scripts more portable.
> 
> It looks like using `ts-node --script-mode` would be a good option as you could have this setup, what do you think?

Hmm. I don't think I want to have a separate `tsconfig.json` anymore, but instead I'd want to prisma to properly use my main `tsconfig.json` in my project root properly respecting the `baseUrl` and `paths`. Although this could be an interesting feature, I don't think it solves my particular problem. Yep I just run my file with `ts-node --script-mode` and I get same error with absolute paths: `Error: Cannot find module 'lib/prisma/prisma'`. Again this fixes it `ts-node -r tsconfig-paths/register ...`.

> Note: `process.exit(0)` should be used for exiting with success and all non zero code are for exiting with an error.

Thanks, I forgot.

> Are you using that [seed template](https://www.prisma.io/docs/guides/prisma-guides/seed-database#integrated-seeding-with-prisma-migrate)?

I was missing that `process.exit(0)`. Here it is https://gist.github.com/JakubKoralewski/061af1318525b59428d44a146168fee9, I've added the ability to run on the production db with an argument, and load the `.env` file, also I am making a prompt with a timeout if I wanted to change the user id. Rest is similar I think.

**Comment by rajuashok** [comment]
> > Changing my `tsconfig.json` to have this:
> > `"module": "commonjs",`
> > instead of this:
> > `"module": "esnext",`
> > fixed it.
> 
> what are the implications of your change? I also get that `seed.ts` error

My app builds fine. I posted about this solution here: https://github.com/TypeStrong/ts-node/issues/922#issuecomment-765008102

This article provides a better understanding of what the module piece means for bundling: https://www.tsmean.com/articles/learn-typescript/typescript-module-compiler-option/.

Hope that helps!

**Comment by keithkraker** [comment]
I ran into a problem with the new seeding feature. My seed data provides hardcoded id values and there's a collision when I try to create new records. My id value is set to `@id @default(autoincrement())` and when I create new records after the seeding process, it tries to add ids that are taken. Is this how it is supposed to work? I thought `@id @default(autoincrement())` would look at the current id values and increment from there. Is there an issue here?

**Comment by albertoperdomo** [comment]
@keithkraker I don't think this is an issue of the seeding feature. I think it's rather an inherent problem of what you are trying to do.

I think for serial primary keys, the database usually generates the IDs based on a sequence (think counter), which probably starts at 1 by default. It won't know the IDs are in use and skip them if that's the case because it relies on the sequence number.

In general, I don't think it's considered good practice to insert hardcoded IDs in tables with serial primary keys. It would seem more reasonable to let the database generate those IDs instead.

If this is a hard requirement, I guess you could reset the sequence number manually in a database migration or something along those lines to make sure any future generated ID does not collide with your hardcoded IDS.

**Comment by keithkraker** [comment]
@albertoperdomo I can remove the id from my seed data and it works fine. I wanted to bring it up just in case it was an error with the new seeding feature or `@id @default(autoincrement())`

**Comment by bigen1925** [comment]
Hi, 

I encountered a problem, and solved it today, so want to share to somebody with a same problem.


I made a seed file following [a guide](https://www.prisma.io/docs/guides/prisma-guides/seed-database) , and executed seeding.
Then, I got an error below.

```
# npx prisma db seed --preview-feature

Running `ts-node seed.ts` ...
TypeError: Class constructor AsyncResource cannot be invoked without 'new'
    at new RequestHandler (/app/prisma/client/runtime/index.js:25399:9)
    at class_2.request (/app/prisma/client/runtime/index.js:25481:27)
    at /app/prisma/client/runtime/index.js:26249:19
    at new Promise (<anonymous>)
    at class_3.request (/app/prisma/client/runtime/index.js:26248:14)
    at class_4.<anonymous> (/app/prisma/client/runtime/index.js:26969:48)
    at step (/app/prisma/client/runtime/index.js:57:23)
    at Object.next (/app/prisma/client/runtime/index.js:38:53)
    at fulfilled (/app/prisma/client/runtime/index.js:29:58)
    at processTicksAndRejections (node:internal/process/task_queues:94:5) {
  clientVersion: '2.15.0'
}
Error: Command failed with exit code 1: ts-node /app/prisma/seed.ts
npm ERR! code 1
npm ERR! path /app
npm ERR! command failed
npm ERR! command sh -c prisma "db" "seed" "--preview-feature"

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-01-31T23_48_25_768Z-debug.log
```

This is a problem of a syntax of Javascript, and I solved this with replacing a `tsconfig.json` option

from
```
{
    "compilerOptions": {
        "target": "es5",
    }
}
```

to
```
{
    "compilerOptions": {
        "target": "es6",
    }
}
```

When seeding is moved to stable features, possibly this requirement should be documented.

I couldn't find a requirement of a compile target in Prisma's document.

Thank you, and I love Prisma ;)

**Comment by adarnon** [comment]
I think there should be some way to control the arguments that Prisma passed to `seed.ts`. My `tsconfig` uses path alises (with `compilerOptions.paths`), so `ts-node` needs to receive `-r tsconfig-paths/register`, and there is no way to control that at the moment.

**Comment by timsuchanek** [comment]
Thanks for reporting @bigen1925! I tried to reproduce the `es5` issue, but couldn't https://github.com/timsuchanek/seeding
Could you share a reproduction? That would be very very awesome!

**Comment by bigen1925** [comment]
@timsuchanek 

I finally could reproduce here https://github.com/timsuchanek/seeding/pull/1/files , but it was somewhat different what I thought.

`target` option is not a direct cause, but `allowJs` option and prisma schema's `output` are.

### Steps of reproduction
- set `tsconfig.json`
```
{
    "compilerOptions": {
        "allowJs": true
    }
}
```

- set `schema.prisma`
```
generator client {
  provider = "prisma-client-js"
  output   = "./client"
}
```

- make a seed file
```js
// import { PrismaClient } from '@prisma/client'
import { PrismaClient } from './client'

async function main() {
  const prisma = new PrismaClient()
  const u = await prisma.user.findFirst()
  console.log(u)
}

main()
```

- run seeding, and it will crash.
```
$ npx prisma db push --preview-feature
$ npx prisma db seed --preview-feature

Environment variables loaded from prisma/.env
Running ts-node "prisma/seed.ts" ...

/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:25400
        super("UNDICI_REQUEST");
        ^
TypeError: Class constructor AsyncResource cannot be invoked without 'new'
    at new RequestHandler (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:25400:9)
    at class_2.request (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:25482:27)
    at /Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:26252:19
    at new Promise (<anonymous>)
    at class_3.request (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:26251:14)
    at class_4.<anonymous> (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:26967:48)
    at step (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:56:23)
    at Object.next (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:37:53)
    at fulfilled (/Users/mitsuru.kasai/Documents/git/seeding/prisma/client/runtime/index.js:28:58)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
Error: Command failed with exit code 1: ts-node "prisma/seed.ts"
```

### FYI
if you add `target: "es6"` option to `tsconfig.json`, seeding works.
```
{
  "compilerOptions": {
    "target": "es6",
    "allowJs": true
  }
}
```

**Comment by flybayer** [comment]
I really do not like the decision to make this `seed.ts` file a script with side effects (vs exporting a function that prisma would then run).

This totally breaks the use case of being able to export a function from the seed file that can be manually invoked somewhere else in my code, such as some test scripts, production scripts, etc.

I urge you to consider changing the functionality to instead export a function so the seed file is side-effect free.

```diff
import { PrismaClient } from '@prisma/client'

async function main() {
  const prisma = new PrismaClient()
  const u = await prisma.user.findFirst()
  console.log(u)
}

-main()
+export default main
```

**Comment by JakubKoralewski** [comment]
> This totally breaks the use case of being able to export a function from the seed file that can be manually invoked somewhere else in my code, such as some test scripts, production scripts, etc.

I totally do this with `if (require.main === module) {` to call the function if it is run explicitly, and the main function is a `default export` in my case.

**Comment by Jolg42** [comment]
It seems doing what @JakubKoralewski is a good solution to me, full example:

```ts
import { PrismaClient } from "@prisma/client"

async function main() {
  const prisma = new PrismaClient()
  // seed
}

if (require.main === module) {
  main()
}

export default main
```

**Comment by flybayer** [comment]
@Jolg42 imo that's very confusing. Most users won't know what the heck is going on there 🤷‍♂️

**Comment by Jolg42** [comment]
@flybayer What is very confusing? `if (require.main === module)`?

Where do you import the seed? I'm trying to understand the full use-case with an example 👀

**Comment by flybayer** [comment]
> @flybayer What is very confusing? `if (require.main === module)`?

Yes, that :)

> Where do you import the seed? I'm trying to understand the full use-case with an example 👀

Currently I'm importing the seed function into tests. Because I'm clearing all the data from the DB in between tests and then need to seed it (I'm not using prisma migrate reset here because I don't need the whole DB destroyed and re-created)

**Comment by fnky** [comment]
I'm with @adarnon with regards to allowing passing options to ts-node. I think it would make sense to follow best practices and allow a way to forward any arguments to ts-node.

The POSIX convention is using a terminator (`--`) which indicates that options passed to the original program have terminated, and anything after can be used by the program to pass to external programs.

```sh
$ prisma migrate dev --preview-feature -- --project tsconfig.prisma.json --compiler-options '{"module":"CommonJS"}'
#                    ^                 ^  ^
#                    |                 |   ` args passed to ts-node
#               prisma args       terminator
```

I think this should be done for all commands that spawns other processes/runs other tools.

I don't know what the standard is in Windows, but since the current CLI tools mostly follow some POSIX-like conventions anyway, I don't think this would be an issue.

**Comment by elkevinwolf** [comment]
I made a temporary workaround while this gets fixed, in case anyone finds it useful:

### 1. Install required dependencies

```
$ npm install -D ts-node tsconfig-paths
```

### 2. Rename `prisma/seed.ts` to `prisma/_seed.ts`

```
$ mv prisma/seed.ts prisma/_seed.ts
```

### 3. Create `seed.sh` and call `ts-node`

```sh
# @TODO: remove this and rename _seed.ts to seed.ts once https://github.com/prisma/prisma/issues/5161
# has been resolved.
npx ts-node --compiler-options '{"module":"CommonJS"}' -r tsconfig-paths/register prisma/_seed.ts
```

**Comment by Jolg42** [comment]
@flybayer We just shipped an improvement to `prisma db seed --preview-feature`

Curious to get your feedback there when you can try it out 😉 

From Release Notes: https://github.com/prisma/prisma/releases/tag/2.23.0
--
In previous versions, a _seed file_ could only be executed as a _script_. `prisma db seed` was simply executing the script by either calling `node ./prisma/seed.js` for JavaScript or `ts-node ./prisma/seed.ts` for TypeScript.

Now, you can directly export a function that Prisma executes on your behalf. Here's the rules that describe the new behavior: Prisma checks if the seed file has ...

- ... an exported function named `seed` and executes it
- ... an exported function (via a _default_ export) and executes it

If there's no function exported as `seed` or via a _default_ export, the behaviour of `prisma db seed` is exactly the same as before, meaning it will simply be executed as a script file.

**Comment by dangreaves** [comment]
@Jolg42 interestingly, I cannot seem to get the exported function to execute on `prisma@2.23.0`. I have included an example below which console logs inside and outside the function to demonstrate.

This is in a default Next.js project.

I have tested with `ts-node@10` and `ts-node@9.1.1`.

## prisma/seed.ts (with named export)

```ts
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

console.log("outer");

export async function seed() {
  console.log("inner");
}
```

## prisma/seed.ts (with default export)

```ts
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

console.log("outer");

async function seed() {
  console.log("inner");
}

export default seed;
```

## tsconfig.json

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": false,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve"
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
  "exclude": ["node_modules"]
}
```

## package.json "ts-node" script

```
"ts-node": "ts-node --compiler-options '{\"module\":\"commonjs\"}'"
```

## `npx prisma db seed --preview-feature` output

This is the same output for both named and default exports.

```sh
$ npx prisma db seed --preview-feature
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Running seed: ts-node --compiler-options '{"module":"commonjs"}' "prisma/seed.ts" ...
outer

🌱  Your database has been seeded.
```

**Comment by blazestudios23** [comment]
I'm using next.js and trying to seed, when I run the commend it doesn't work. But every time I start next.js it seeds the db, but causes errors due to unique constraints. 

This is extremely inconvenient.

**Comment by janpio** [comment]
Please open a new issue @blazestudios23 and provide some information about your project so we can possibly reproduce and understand this. Thanks.

**Comment by JohnLawliet** [comment]
I'm using prisma 2.24.1 and `npx prisma db seed --preview-feature` is not working in .js file. I have the seed.js in my prisma folder and have tried out import/export by adding `{ "type": "module" }` but it still just says 
"Your database has been seeded."
without actually seeding the database. Please fix seeding for .js files.

**Comment by janpio** [comment]
Are you on Windows @JohnLawliet? Then this is covered (and being worked on) in https://github.com/prisma/prisma/issues/7176
If not, optimally open a new issue so we can investigate. Make sure to fill the issue template so we have all information we need.

**Comment by leerob** [comment]
Please let me know if there's a corresponding issue on the Next.js side related to this @janpio 🙏

**Comment by janpio** [comment]
Reminder @blazestudios23 - is that problem you mentioned above still present? We (and Vercel, the people behind Next.js, see @leerob's message) would love to understand the problem and be able to fix it. Right now I can not find an issue about this yet, so we are having trouble understanding and reproducing what is going on for you.

**Comment by nghiepit** [comment]
How to seeding with the specific object to run individually?
For example to create seed file and run specific:

```bash
$ npx prisma db seed --create=UserSeed
$ npx prisma db seed --class=UserSeed
```

**Comment by benbender** [comment]
To use `prisma db seed --preview-feature` in a native es-module project (`type: module` in package.json), I found the following solution:

1. `npm add ts-node`
2. add `"ts-node": "node --loader ts-node/esm"` to `scripts` in package.json`
3. use a `prisma/seed.ts` similar to this:

````ts
import prisma from '@prisma/client';

const { PrismaClient } = prisma;

export const seed = async () => {
	const db = new PrismaClient();

	// Prisma create query to seed models in database
	const alice = await db.user.upsert({
		where: { email: 'alice@prisma.io' },
		update: {},
		create: {
			email: 'alice@prisma.io',
			name: 'Alice',
			hashedPassword: 'foobar',
			posts: {
				create: {
					title: 'Check out Prisma with Next.js',
					content: 'https://www.prisma.io/nextjs'
					// published: true
				}
			}
		}
	});

	const bob = await db.user.upsert({
		where: { email: 'bob@prisma.io' },
		update: {},
		create: {
			email: 'bob@prisma.io',
			name: 'Bob',
			hashedPassword: 'foobar',
			posts: {
				create: [
					{
						title: 'Follow Prisma on Twitter',
						content: 'https://twitter.com/prisma'
						// published: true
					},
					{
						title: 'Follow Nexus on Twitter',
						content: 'https://twitter.com/nexusgql'
						// published: true
					}
				]
			}
		}
	});

	console.log({ alice, bob });

	await db.$disconnect();
};
````

**Comment by benbender** [comment]
Hey, thanks for this feature. Here are my 

feedback/Wishlist:
 [ ] define one clear way how the seed-file is supposed to be called (instead of multiple options of exporting it)
 [ ] add a way to customize the seed-runner-script as there may be esoteric setups which may have problems with yours
 [ ] programatic api to use the seed-runner from scripts (f.e. tests)
 [ ] handle db-connection-setup/teardown in the runner and inject it into the seed-file
 [ ] ability to have a directory with individual seeders per model
 [ ] ability to run individual model-seed-files (f.e. nice for testing)

**Comment by Jolg42** [comment]
Hi @benbender Thank you for the suggestions!

We are reworking the seeding feature at the moment:
The gist of it is that we want seeding to always work like that ts-node escape hatch, but with a clearer naming and guidance in the CLI.

Would you be interested in testing a preview version of the feature and giving us your feedback? (that offer applies to anyone reading this message).

**Comment by benbender** [comment]
@Jolg42 sure, why not? :) anything I need to do?

**Comment by AlexKMarshall** [comment]
I am seeing something similar to @blazestudios23 - in a NextJS project, using sqlite, my seed function (a named export) is not getting called.

I've created an issue https://github.com/prisma/prisma/issues/8685 for that.

I have a repo branch demonstrating the issue https://github.com/AlexKMarshall/next-prisma-starter/tree/prisma-seeds

Hope that that helps in tracking down what's going on. I'll try running it as a script file, rather than a function, but it would be very useful to be able to call it from tests too, so I'd definitely like to get the function working if possible.

Thanks

**Comment by gex** [comment]
i'm using the the `seed` command with the newer named export syntax but the process doesn't exit.

`seed.ts`:
```ts
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

export async function seed() {
  const user = await prisma.user.upsert({
    create: {
      name: 'User,
      email: 'user@example.com',
    },
    update: {},
    where: {
      email: 'user@example.com',
    },
  })

  console.log({ user })
}
```

this is pretty much the same as [in the example](https://www.prisma.io/docs/guides/database/seed-database#seeding-your-database-with-typescript). everything seems fine (record is created in the db), except it just hangs and doesn't exit at the end. tried `prisma db seed --preview-feature` and `prisma migrate reset` as well, with the same result.

prisma version and env:
```
prisma               : 2.29.0
@prisma/client       : 2.29.0
Current platform     : linux-musl
Query Engine         : query-engine 1be4cd60b89afa04b192acb1ef47758a39810f3a (at node_modules/@prisma/engines/query-engine-linux-musl)
Migration Engine     : migration-engine-cli 1be4cd60b89afa04b192acb1ef47758a39810f3a (at node_modules/@prisma/engines/migration-engine-linux-musl)
Introspection Engine : introspection-core 1be4cd60b89afa04b192acb1ef47758a39810f3a (at node_modules/@prisma/engines/introspection-engine-linux-musl)
Format Binary        : prisma-fmt 1be4cd60b89afa04b192acb1ef47758a39810f3a (at node_modules/@prisma/engines/prisma-fmt-linux-musl)
Default Engines Hash : 1be4cd60b89afa04b192acb1ef47758a39810f3a
Studio               : 0.419.0
```

**Comment by Jolg42** [comment]
Hi @gex so you will need to call `await prisma.$disconnect()` at the end of your seed function. That should work.

**Comment by Jolg42** [comment]
🗒️ We heard the feedback and we are planning a redesign of `db seed` (currently planned to be released early September)

How it will work is that a `prisma.seed` property in the package.json of your project will be required if you want to use `prisma db seed`.

This property is where you will be able to put any command you want to execute your seed script.

More info/details & how to try it out now in:
https://github.com/prisma/prisma/issues/8732

**Comment by gex** [comment]
thanks @Jolg42 ! it seems like my "pretty much the same as in the example implementation" missed a key part 😅 i read @benbender 's idea of automatically setup the db connection first and i think script vs function and that idea added up in my head and assumed it's already working like that. long story short it would be definitely awesome of having the db connection passed to the exported seed function automatically. otherwise it's just confusing that there is a script and a function syntax. or does #8732 mean the latter won't work anymore?

**Comment by 2color** [comment]
Closing this now that this has been released to General Availability in the [3.0.1 release](https://github.com/prisma/prisma/releases/tag/3.0.1)
